lib/mayaUsd/fileio/shaderReaderRegistry.cpp

```diff
@@ -40,51 +40,98 @@ TF_DEFINE_PRIVATE_TOKENS(
+
+_Registry::const_iterator _Find(
+    const TfToken&              usdInfoId,
+    const UsdMayaJobImportArgs& importArgs) {
+    using ContextSupport = UsdMayaShaderReader::ContextSupport;
+
+    _Registry::const_iterator ret = _reg.cend();
+    _Registry::const_iterator first, last;
+    std::tie(first, last) = _reg.equal_range(usdInfoId);
+    while (first != last) {
+        ContextSupport support = first->second._pred(importArgs);
+        if (support == ContextSupport::Supported) {
+            ret = first;
+        } else if (support == ContextSupport::Fallback && ret == _reg.end()) {
+            ret = first;
+        }
+        ++first;
+    }
+
+    return ret;
+}
+} // namespace
 
 /* static */
 void UsdMayaShaderReaderRegistry::Register(
-    TfToken                                      usdInfoId,
-    UsdMayaShaderReaderRegistry::ReaderFactoryFn fn)
+    TfToken                                         usdInfoId,
+    UsdMayaShaderReaderRegistry::ContextPredicateFn pred,
+    UsdMayaShaderReaderRegistry::ReaderFactoryFn    fn)
 {
     TF_DEBUG(PXRUSDMAYA_REGISTRY)
         .Msg("Registering UsdMayaShaderReader for info:id %s.\n", usdInfoId.GetText());
 
-    std::pair<_Registry::iterator, bool> insertStatus = _reg.insert(std::make_pair(usdInfoId, fn));
-    if (insertStatus.second) {
-        UsdMaya_RegistryHelper::AddUnloader([usdInfoId]() { _reg.erase(usdInfoId); });
-    } else {
-        TF_CODING_ERROR("Multiple readers for id %s", usdInfoId.GetText());
-    }
+    int index = _indexCounter++;
+    _reg.insert(std::make_pair(usdInfoId, _RegistryEntry{pred, fn, index}));
+
+    // The unloader uses the index to know which entry to erase when there are
+    // more than one for the same usdInfoId.
+    UsdMaya_RegistryHelper::AddUnloader([usdInfoId, index]() { 
+        _Registry::const_iterator it, itEnd;
+        std::tie(it, itEnd) = _reg.equal_range(usdInfoId);
+        for (; it != itEnd; ++it) {
+            if (it->second._index == index) {
+                _reg.erase(it);
+                break;
+            }
+        }
+    });
 }
 
 /* static */
-UsdMayaShaderReaderRegistry::ReaderFactoryFn
-UsdMayaShaderReaderRegistry::Find(const TfToken& usdInfoId)
+UsdMayaShaderReaderRegistry::ReaderFactoryFn UsdMayaShaderReaderRegistry::Find(
+    const TfToken&              usdInfoId,
+    const UsdMayaJobImportArgs& importArgs)
 {
+    using ContextSupport = UsdMayaShaderReader::ContextSupport;
     TfRegistryManager::GetInstance().SubscribeTo<UsdMayaShaderReaderRegistry>();
 
-    ReaderFactoryFn ret = nullptr;
-    if (TfMapLookup(_reg, usdInfoId, &ret)) {
-        return ret;
+    _Registry::const_iterator it = _Find(usdInfoId, importArgs);
+    
+    if (it != _reg.end()) {
+        return it->second._fn;
     }
 
+    // Try adding more writers via plugin load:
     static const TfTokenVector SCOPE = { _tokens->UsdMaya, _tokens->ShaderReader };
     UsdMaya_RegistryHelper::FindAndLoadMayaPlug(SCOPE, usdInfoId);
```

Comment:
It is a pattern I saw used in all of the Prim{reader|updater|writer} registries, where a double TfMapLookup is done, with a nullptr dummy used to prevent loading plugins twice. It is replicated here in expanded form, but that is due to unordered_multimap usage.